[Streams 🌊] Warn about outdated documents

AlejandroFrndz opened the pull request:
## Summary

In order to have a better change tracking system for streams, this PR introduces `updated_at` timestamps for streams documents that are meant to record the moment in time when changes occur in a stream. There are 2 of this timestamps being introduced. Each one has a different level of nesting, given that they track changes to specific parts of the whole stream document.

The first one lives at the root of the stream document (next to name and description) and tracks changes to any part of the stream, which means all upsert requests processed for the stream will update it. It is present in all stream types.

The second one is more specific and only present for ingest streams (Classic and Wired). It lives in `ingest.processing.updated_at` and tracks changes exclusively for the processing config of the stream. This property will be used in the future to display outdated document warnings in the streams UI while working in the processing tab.

The processing timestamp is then used by the UI to determine wether to show an outdated documents warning when configuring the processing step. The warning will be shown when at least 1 document in the current simulation samples set is older than the most recent stream processing timestamp. 

- The warning will prompt users to refresh the samples in order to get up to date data. But, it may be possible for the warning not to go away after refreshing if ingestion is stopped or slow
- The outdated documents logic only applies to latest and KQL samples. It will never be displayed for custom samples sets given that the dataset in this case is static thus the concept of being outdated doesn't really apply.

### Visual Updates 👀

<img width="1365" height="907" alt="Screenshot 2025-11-24 at 12 05 55" src="https://github.com/user-attachments/assets/405ff623-339c-49ce-aefe-5386cfd9030d" />

### Reviewer hints

The most important changes are in the `kbn-streams-schema` package, mostly around ingest types and model_validation. So please pay a closer look at the changes there

Other significant changes are in the `streams` plugin,  under `state_management` (where the decision to update the processing timestamp is made) and `migrate_on_read.ts` to backfill the new properties in existing streams.

For the UI, the most relevant updates are around the state machines for data sources and stream processing. The data source machine is responsible for determining whether the warning should be shown or not, as those machines are the owners of the data in the samples and also know the kind of data source that also influences wether the icon should show.
The stream processing machine also plays its part. Given that it owns the stream definition, it handles forwarding the latest processing timestamp to the data source machines via machine-to-machine events

## `terminate_after` Updates
To address the issue detected in https://github.com/elastic/kibana/pull/242904#pullrequestreview-3515851161 we have decided to remove the `terminate_after` parameter from the queries we run to fetch documents for the processing simulation. This will ensure that data sources, other than custom samples, always return the most up to date samples from the stream.

This could potentially cause performance issues in samples documents retrieval, given that we no longer have the early termination in shards that `terminate_after` was providing. Even though queries for samples might now be slower, we consider that a worthy trade-off in exchange for getting the most up to date documents and fixing the ever present warning.

### Telemetry
 A new custom EBT event was added, `streams-processing-simulation-samples-fetch-latency` that will report statistics on the execution time of simulations samples queries. This way we can keep an eye out on this and detect wether the query time is too high.

### Timeout
To prevent the page from hanging indefinitely while waiting for the query to run, we've added a 10s timeout in the data collector actor for simulations samples on ES side. If any query take longer than that we will do a best effort approach to display any samples returned as a partial result from ES before the query timed out

AlejandroFrndz committed:
Add updated_at to stream definitions and migrate missing values

Adds an updated_at timestamp to stream definitions and ensures missing values are populated on read (defaults to the epoch). Updates validation to require a datetime string and tightens TypeScript types for upsert requests so clients cannot set name or updated_at.

This prevents schema/validation gaps for older data, normalizes stream metadata across wired, classic and group streams, and includes unit tests covering migration and preservation of existing timestamps.

AlejandroFrndz committed:
Rejects updated_at in UpsertRequest and updates tests

Adds validation to disallow clients from setting updated_at on upsert payloads while keeping internal definitions satisfied.

Injects a default updated_at when coercing the stream definition for internal validation and then removes it so updated_at remains server-controlled. Updates unit tests to include updated_at where appropriate and adds negative cases to ensure updated_at is rejected on UpsertRequest inputs (plus extra edge-case tests for missing/invalid fields).

AlejandroFrndz committed:
fix: some more jest tests

AlejandroFrndz committed:
fix: Fix type violations

kibanamachine committed:
Changes from node scripts/capture_oas_snapshot   --include-path /api/status   --include-path /api/alerting/rule/   --include-path /api/alerting/rules   --include-path /api/actions   --include-path /api/security/role   --include-path /api/spaces   --include-path /api/streams   --include-path /api/fleet   --include-path /api/saved_objects/_import   --include-path /api/saved_objects/_export   --include-path /api/maintenance_window   --include-path /api/agent_builder --update

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/361080)
* Commit: a7c7306dcaa3e58240a987cdec69264224a2e38c

### Failed CI Steps
* [Check OAS Snapshot](https://buildkite.com/elastic/kibana-pull-request/builds/361080#019a7dea-6e4f-4463-bd0a-673e7aded187)

cc @AlejandroFrndz

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361080","number":361080,"commit":"a7c7306dcaa3e58240a987cdec69264224a2e38c"}],"number":361080}
buildkite-pr-comment-->

kibanamachine committed:
Changes from make api-docs

AlejandroFrndz committed:
Adds updated_at and description tests; make timestamp expectation deterministic

Adds focused tests that validate stream description setting/updating and verify updated_at behavior for new and existing streams. Organizes metadata tests into descriptive suites to improve clarity and coverage. Also changes expected updated_at in migration tests to a fixed epoch timestamp to remove time-dependent flakiness and make expectations deterministic.

AlejandroFrndz committed:
commit with issues

AlejandroFrndz committed:
potential fix

AlejandroFrndz committed:
first working version for scoped upsert

AlejandroFrndz committed:
integrate processing updated at

AlejandroFrndz committed:
Add ingest upsert schemas

Introduce upsert-specific validation for ingest streams so clients can send processing without an updated_at timestamp.

Updates the processing schema to merge the DSL with updated_at and adds Upsert request types that omit required processing.updated_at. Adjusts server route, UI upsert actor, and tests to use the new schemas and stop manually injecting or stripping updated_at. Allows updated_at to be managed server-side and simplifies the client upsert flow.

AlejandroFrndz committed:
some type fixes

AlejandroFrndz committed:
Types naming cleanup

kibanamachine committed:
Changes from node scripts/capture_oas_snapshot   --include-path /api/status   --include-path /api/alerting/rule/   --include-path /api/alerting/rules   --include-path /api/actions   --include-path /api/security/role   --include-path /api/spaces   --include-path /api/streams   --include-path /api/fleet   --include-path /api/saved_objects/_import   --include-path /api/saved_objects/_export   --include-path /api/maintenance_window   --include-path /api/agent_builder --update

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/362442)
* Commit: 579eedddc5f17f65c41344459aac02eb2a52b18f

### Failed CI Steps
* [Check OAS Snapshot](https://buildkite.com/elastic/kibana-pull-request/builds/362442#019a9860-3359-437f-ada6-1b070b27c18f)

### History
* :green_heart: [Build #361117](https://buildkite.com/elastic/kibana-pull-request/builds/361117) succeeded 8bcc6e3124d03b1214df5f7d3ca96faa1da51604
* :broken_heart: [Build #361099](https://buildkite.com/elastic/kibana-pull-request/builds/361099) failed c5a5f2bc037a55a73f907a96abb07be672b2a1f3
* :broken_heart: [Build #361080](https://buildkite.com/elastic/kibana-pull-request/builds/361080) failed a7c7306dcaa3e58240a987cdec69264224a2e38c

cc @AlejandroFrndz

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362442","number":362442,"commit":"579eedddc5f17f65c41344459aac02eb2a52b18f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361117","number":361117,"commit":"8bcc6e3124d03b1214df5f7d3ca96faa1da51604"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361099","number":361099,"commit":"c5a5f2bc037a55a73f907a96abb07be672b2a1f3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361080","number":361080,"commit":"a7c7306dcaa3e58240a987cdec69264224a2e38c"}],"number":362442}
buildkite-pr-comment-->

kibanamachine committed:
Changes from make api-docs

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/362446)
* Commit: 7f1baef235cccb2c1313657b50335d3fd7b356ef

### Failed CI Steps
* [Check OAS Snapshot](https://buildkite.com/elastic/kibana-pull-request/builds/362446#019a9875-7cbf-4453-bda8-3b8c7bcbb0b6)

### History
* :broken_heart: [Build #362442](https://buildkite.com/elastic/kibana-pull-request/builds/362442) failed 579eedddc5f17f65c41344459aac02eb2a52b18f
* :green_heart: [Build #361117](https://buildkite.com/elastic/kibana-pull-request/builds/361117) succeeded 8bcc6e3124d03b1214df5f7d3ca96faa1da51604
* :broken_heart: [Build #361099](https://buildkite.com/elastic/kibana-pull-request/builds/361099) failed c5a5f2bc037a55a73f907a96abb07be672b2a1f3
* :broken_heart: [Build #361080](https://buildkite.com/elastic/kibana-pull-request/builds/361080) failed a7c7306dcaa3e58240a987cdec69264224a2e38c

cc @AlejandroFrndz

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362446","number":362446,"commit":"7f1baef235cccb2c1313657b50335d3fd7b356ef"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362442","number":362442,"commit":"579eedddc5f17f65c41344459aac02eb2a52b18f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361117","number":361117,"commit":"8bcc6e3124d03b1214df5f7d3ca96faa1da51604"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361099","number":361099,"commit":"c5a5f2bc037a55a73f907a96abb07be672b2a1f3"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361080","number":361080,"commit":"a7c7306dcaa3e58240a987cdec69264224a2e38c"}],"number":362446}
buildkite-pr-comment-->

AlejandroFrndz committed:
Consolidate stream conversion helpers

Merge and centralize stream conversion logic into a single converters module and rename the get→upsert helper for clarity.

This consolidates two conversion helpers into one file, removes the deprecated helper, and moves/updates tests to validate both directions. UI code imports are updated to use the new, clearer name for converting a GetResponse into an UpsertRequest. Benefits include reduced duplication, clearer naming, and easier maintenance of conversion logic.

AlejandroFrndz committed:
Add IngestUpsertRequest validation tests

Adds validation tests for ingest upsert requests for classic and wired stream models. The tests cover valid payload shapes and several invalid cases (missing required sections and disallowed updated_at in processing), improving schema coverage and preventing regressions in request parsing/validation.

AlejandroFrndz committed:
Adds comment clarifying updated_at migration

Adds a clarifying comment that documents adding an `updated_at` field to the `processing` object for ingest streams during migration-on-read.

Aids readability and future maintenance by explaining the intent of the subsequent migration logic. No functional behavior is changed.

AlejandroFrndz committed:
fix synthtrace scenario

AlejandroFrndz committed:
Fix scout fixture

AlejandroFrndz committed:
documentation fixes

AlejandroFrndz committed:
Update model validation comment

fake-haris reviewed (approved):
src/platform/packages/shared/kbn-scout/src/playwright/fixtures/scope/worker/apis/streams/index.ts changes LGTM

AlejandroFrndz committed:
Fix test spec

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/362682)
* Commit: 6c76e28227f19bdac3a81fa7f1a251895a04e4d8
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-242904)

### Failed CI Steps
* [Scout: [ platform / streams_app ] plugin](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c7c-a66d-4e51-972e-b2817e29475c)
* [Scout: [ platform / streams_app ] plugin](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c9c-4e07-47d3-bd91-e861013f0039)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c7c-a66d-4e51-972e-b2817e29475c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/362682/jobs/019a9c7c-a66d-4e51-972e-b2817e29475c/artifacts/019a9c9c-1be9-41f9-929e-0b67bf5ef7d2) Scout: [ platform / streams_app ] plugin / serverless-oblt - Stream data routing - editing routing rules - should edit an existing routing rule
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c9c-4e07-47d3-bd91-e861013f0039) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/362682/jobs/019a9c9c-4e07-47d3-bd91-e861013f0039/artifacts/019a9cbc-1236-4572-b498-dc5a67d27696) Scout: [ platform / streams_app ] plugin / serverless-oblt - Stream data routing - editing routing rules - should edit an existing routing rule
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c7c-a66d-4e51-972e-b2817e29475c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/362682/jobs/019a9c7c-a66d-4e51-972e-b2817e29475c/artifacts/019a9c9c-1bea-4060-8f05-654cf73d6f16) Scout: [ platform / streams_app ] plugin / serverless-oblt - Stream data routing - error handling and recovery - should recover from API errors during rule updates
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c9c-4e07-47d3-bd91-e861013f0039) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/362682/jobs/019a9c9c-4e07-47d3-bd91-e861013f0039/artifacts/019a9cbc-1238-4b8f-8654-04528306297e) Scout: [ platform / streams_app ] plugin / serverless-oblt - Stream data routing - error handling and recovery - should recover from API errors during rule updates
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c7c-a66d-4e51-972e-b2817e29475c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/362682/jobs/019a9c7c-a66d-4e51-972e-b2817e29475c/artifacts/019a9c9c-1bea-4455-97d9-7805737ad3dc) Scout: [ platform / streams_app ] plugin / should edit an existing routing rule
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c9c-4e07-47d3-bd91-e861013f0039) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/362682/jobs/019a9c9c-4e07-47d3-bd91-e861013f0039/artifacts/019a9cbc-1237-4e3b-87a5-bd3bcd856c8a) Scout: [ platform / streams_app ] plugin / should edit an existing routing rule
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c7c-a66d-4e51-972e-b2817e29475c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/362682/jobs/019a9c7c-a66d-4e51-972e-b2817e29475c/artifacts/019a9c9c-1bea-4ed4-8d69-9f51f4d98ad1) Scout: [ platform / streams_app ] plugin / should recover from API errors during rule updates
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c9c-4e07-47d3-bd91-e861013f0039) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/362682/jobs/019a9c9c-4e07-47d3-bd91-e861013f0039/artifacts/019a9cbc-1238-4fb8-85c9-0d9b2406e27a) Scout: [ platform / streams_app ] plugin / should recover from API errors during rule updates
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c7c-a66d-4e51-972e-b2817e29475c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/362682/jobs/019a9c7c-a66d-4e51-972e-b2817e29475c/artifacts/019a9c9c-1be9-430e-bcf2-03eebf527116) Scout: [ platform / streams_app ] plugin / stateful - Stream data routing - editing routing rules - should edit an existing routing rule
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c9c-4e07-47d3-bd91-e861013f0039) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/362682/jobs/019a9c9c-4e07-47d3-bd91-e861013f0039/artifacts/019a9cbc-1236-47d0-8abf-c71261847781) Scout: [ platform / streams_app ] plugin / stateful - Stream data routing - editing routing rules - should edit an existing routing rule
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c7c-a66d-4e51-972e-b2817e29475c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/362682/jobs/019a9c7c-a66d-4e51-972e-b2817e29475c/artifacts/019a9c9c-1be9-472e-9a29-f4a635158197) Scout: [ platform / streams_app ] plugin / stateful - Stream data routing - error handling and recovery - should recover from API errors during rule updates
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/362682#019a9c9c-4e07-47d3-bd91-e861013f0039) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/362682/jobs/019a9c9c-4e07-47d3-bd91-e861013f0039/artifacts/019a9cbc-1237-4da5-a007-0b07d270ae9a) Scout: [ platform / streams_app ] plugin / stateful - Stream data routing - error handling and recovery - should recover from API errors during rule updates

### History
* :broken_heart: [Build #362662](https://buildkite.com/elastic/kibana-pull-request/builds/362662) failed 2d00caf9ad605907a758a11fef88633c4ffb1322
* :broken_heart: [Build #362555](https://buildkite.com/elastic/kibana-pull-request/builds/362555) failed 2b1dc80dd1d4a34891af88465749dbc9a8f2b18c
* :broken_heart: [Build #362446](https://buildkite.com/elastic/kibana-pull-request/builds/362446) failed 7f1baef235cccb2c1313657b50335d3fd7b356ef
* :broken_heart: [Build #362442](https://buildkite.com/elastic/kibana-pull-request/builds/362442) failed 579eedddc5f17f65c41344459aac02eb2a52b18f
* :green_heart: [Build #361117](https://buildkite.com/elastic/kibana-pull-request/builds/361117) succeeded 8bcc6e3124d03b1214df5f7d3ca96faa1da51604

cc @AlejandroFrndz

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362682","number":362682,"commit":"6c76e28227f19bdac3a81fa7f1a251895a04e4d8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362662","number":362662,"commit":"2d00caf9ad605907a758a11fef88633c4ffb1322"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362555","number":362555,"commit":"2b1dc80dd1d4a34891af88465749dbc9a8f2b18c"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362446","number":362446,"commit":"7f1baef235cccb2c1313657b50335d3fd7b356ef"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/362442","number":362442,"commit":"579eedddc5f17f65c41344459aac02eb2a52b18f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/361117","number":361117,"commit":"8bcc6e3124d03b1214df5f7d3ca96faa1da51604"}],"number":362682}
buildkite-pr-comment-->

AlejandroFrndz committed:
Fixes processing updated_at preservation logic

Invert the condition that decides when to reuse the existing ingest.processing.updated_at during stream definition upserts. The preservation check now executes in the correct branch so upserted definitions no longer end up with incorrect or stale processing timestamps.

This resolves inconsistent updated_at handling for classic and wired stream flows and prevents misleading timestamps after updates.

rmyz reviewed (approved):
infra & services changes LGTM

flash1293 reviewed (commented):
This is a high level comment, but I wonder whether we should handle these updated_at timestamps outside of the definition - seems to simplify a couple things

AlejandroFrndz committed:
Preserves processing updated_at when unchanged

Ignores the processing.updated_at field when comparing processing definitions so timestamp-only differences do not count as a change. Fixes the copy logic to preserve the previous processing.updated_at when the processing content is actually unchanged rather than when it was detected as changed.

Reduces spurious change detection and timestamp churn on upsert operations.

AlejandroFrndz committed:
Propagates stream timestamp to data sources

Ensures the latest stream ingest.processing.updated_at is passed into each data source actor on spawn and whenever the stream definition updates. Adds an event/action to forward the timestamp from the parent machine and a handler in data source actors to store it. Keeps data sources synchronized with the stream processing state so UI, simulators, and downstream logic use the current ingest timing.

AlejandroFrndz committed:
Adds outdated-document detection for data sources

Converts streamProcessingUpdatedAt from string to Date to allow accurate time comparisons.

Adds a hasOutdatedDocuments flag and a check that determines whether the oldest sample @timestamp precedes the stream's processing updated_at, scoped to sample-based data sources (latest-samples, kql-samples).

Triggers the check whenever data or the processing timestamp updates and propagates the stream processing updated_at from the stream manager to child data source actors.

Improves detection of stale sample previews so UI/logic can surface or handle outdated documents.

AlejandroFrndz committed:
Adds warning icon for outdated simulation samples

Adds detection of outdated documents and renders a warning icon with a tooltip next to the refresh control in the simulation playground. This surfaces when sample data is stale and prompts users to refresh, improving visibility and reducing confusion around outdated simulation results.

flash1293 commented on x-pack/platform/packages/shared/kbn-streams-schema/src/models/validation/model_validation.ts:
```diff
@@ -105,16 +110,35 @@ export function modelValidation(...args: [ModelValidation, ModelSchema] | [Model
+            .transform((prev) => ({
+              ...prev,
+              name: '.',
+              updated_at: new Date().toISOString(),
```
yeah, that's what I had in mind - you would store it by the side and filter it out when turning the source into the definition. It would get a bit uglier on that level for sure, but seems a bit easier to manage - at lot of the stuff you had to touch on your PR was because there is this additional prop in the definition now.

I'll leave it to your and @dgieselaar s descretion though whether it's worth exploring that option.

AlejandroFrndz committed:
Add UI tests and test-id for outdated documents tip

Adds a data-test-subj attribute to the outdated-documents tip anchor to make it selectable in functional tests, and introduces end-to-end UI tests that validate the warning tip visibility.

The tests generate synthetic log streams (new, old, empty) and assert that the warning tip is shown for outdated samples when using latest/KQL sample data, hidden for up-to-date or empty samples, and never shown for custom sample data. This ensures correct behavior and prevents regressions in the outdated-documents warning UX.

AlejandroFrndz committed:
type fixes

AlejandroFrndz committed:
test fixes

flash1293 commented on x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/state_management/data_source_state_machine/data_source_state_machine.ts:
```diff
@@ -82,6 +82,32 @@ export const dataSourceMachine = setup({
+
+      const documentsTimestamps = context.data
+        .map((doc) => doc['@timestamp'])
+        .filter((timestamp): timestamp is string => typeof timestamp === 'string')
```
This won't work for docs like this:
```
{
  "@timestamp": 1764061796490,
  "stream.name": "logs.linux",
  "resource.attributes.host.name": "host1",
  "attributes.process.id": 5678,
  "attributes.filepath": "Linux.log",
  "attributes.user.name": "user3",
  "attributes.kubernetes.pod.name": "pod-303",
  "body.text": "Nov 25 09:09:56 combo sshd(pam_unix)[19937]: check pass; user unknown"
}
```

In general it's tricky to read the time properly from source, it might not be a format that is compatible with `Date` at all.

The way Discover handles this is to request the timestamp via `field` as well, which will return it in a specified format:
```
[
    {
        "field": "@timestamp",
        "format": "date_time"
    }
]
```

not sure whether we need to go this far here, though, we could try to parse the timestamps with the information we have and just disable the feature if it doesn't work

flash1293 reviewed (changes requested):
I think it's OK to keep the updated_at in the definition, it's probably a pain to manage it both ways. However, there are two things we need to fix, left comments in the code

AlejandroFrndz committed:
Parses numeric timestamps with moment

Parses both string and numeric timestamp values using moment and filters out invalid dates before converting to epoch milliseconds.

Improves robustness of outdated-document detection by accepting numeric timestamps and avoiding NaN/invalid date values that could break timestamp calculations.

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/365625)
* Commit: a451949c5674d0a6a73f0b8ed4cb7265387830c9
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-242904)

### Failed CI Steps
* [FTR Configs #60](https://buildkite.com/elastic/kibana-pull-request/builds/365625#019abfcc-9637-4732-a06d-76b41bd0101c)
* [Post-Build](https://buildkite.com/elastic/kibana-pull-request/builds/365625#019abfa7-29ef-4c27-9b73-56c32d7dd845)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ‼️ ERROR: no builds found for mergeBase sha [e6a28149275dd2b7cd1ed024d8ec42dfef40117b]

### History
* :green_heart: [Build #364356](https://buildkite.com/elastic/kibana-pull-request/builds/364356) succeeded d389c7ebaf8360df257d7a15201877d30e1ef184
* :broken_heart: [Build #364239](https://buildkite.com/elastic/kibana-pull-request/builds/364239) failed 0b5fbef274da91ab99fbaec23cc774f1b079a40d
* :broken_heart: [Build #364213](https://buildkite.com/elastic/kibana-pull-request/builds/364213) failed 0b5fbef274da91ab99fbaec23cc774f1b079a40d
* :broken_heart: [Build #364175](https://buildkite.com/elastic/kibana-pull-request/builds/364175) failed b35e3fecfa97aab667acb32b49d701c5fa2a7264
* :green_heart: [Build #363255](https://buildkite.com/elastic/kibana-pull-request/builds/363255) succeeded aab10af5b8df0965c32e692a0fffa2ae514590ee
* :green_heart: [Build #362843](https://buildkite.com/elastic/kibana-pull-request/builds/362843) succeeded 81792e471802e0a96c10789ca5d17ca0ddb8b2dd

cc @AlejandroFrndz

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365625","number":365625,"commit":"a451949c5674d0a6a73f0b8ed4cb7265387830c9"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364356","number":364356,"commit":"d389c7ebaf8360df257d7a15201877d30e1ef184"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364239","number":364239,"commit":"0b5fbef274da91ab99fbaec23cc774f1b079a40d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364213","number":364213,"commit":"0b5fbef274da91ab99fbaec23cc774f1b079a40d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/364175","number":364175,"commit":"b35e3fecfa97aab667acb32b49d701c5fa2a7264"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/363255","number":363255,"commit":"aab10af5b8df0965c32e692a0fffa2ae514590ee"}],"number":365625}
buildkite-pr-comment-->

flash1293 reviewed (commented):
Very impressive work! I went through the code in detail and it mostly looks good to me, just left some smaller notes.

AlejandroFrndz committed:
refactor: Avoid casting in converters

AlejandroFrndz committed:
refactor: Calculate outdated documents via selector

AlejandroFrndz committed:
refactor: Remove machine-to-machine comms for outdated documents

flash1293 reviewed (commented):
Small nit: Layout breaks a bit on small screens

<img width="329" height="129" alt="Image" src="https://github.com/user-attachments/assets/4a240870-4941-4a26-b855-78246449fbfc" />

But that's not a big deal.

The bigger issue is that in practice it's hard to get rid of the warning due to the way we sample.

I kept refreshing on a stream I had and it never went away, even after minutes with active ingest going on. These are the timestamps for the docs:
```
[
    "2025-11-27T14:39:55.063Z",
    "2025-11-27T14:39:53.663Z",
    "2025-11-27T14:39:52.363Z",
    "2025-11-27T14:39:51.063Z",
    "2025-11-27T14:39:49.763Z",
    "2025-11-27T14:39:48.463Z",
    "2025-11-27T14:39:47.163Z",
    "2025-11-27T14:39:45.863Z",
    "2025-11-27T14:39:44.463Z",
    "2025-11-27T14:39:43.163Z",
    "2025-11-27T14:39:41.863Z",
    "2025-11-27T14:39:40.563Z",
    "2025-11-27T14:39:39.263Z",
    "2025-11-27T14:39:37.863Z",
    "2025-11-27T14:39:36.463Z",
    "2025-11-27T14:39:35.163Z",
    "2025-11-27T14:39:33.863Z",
    "2025-11-27T14:39:32.563Z",
    "2025-11-27T14:39:31.163Z",
    "2025-11-27T14:39:29.863Z",
    "2025-11-27T14:39:28.463Z",
    "2025-11-27T14:39:27.163Z",
    "2025-11-27T14:39:25.863Z",
    "2025-11-27T14:39:24.563Z",
    "2025-11-27T14:39:23.163Z",
    "2025-11-27T14:39:21.863Z",
    "2025-11-27T14:39:20.563Z",
    "2025-11-27T14:39:19.163Z",
    "2025-11-27T14:39:17.863Z",
    "2025-11-27T14:39:16.563Z",
    "2025-11-27T14:39:15.263Z",
    "2025-11-27T14:39:13.963Z",
    "2025-11-27T14:39:12.663Z",
    "2025-11-27T14:39:11.263Z",
    "2025-11-27T14:39:09.963Z",
    "2025-11-27T14:39:08.563Z",
    "2025-11-27T14:39:07.263Z",
    "2025-11-27T14:39:05.963Z",
    "2025-11-27T14:39:04.563Z",
    "2025-11-27T14:39:03.063Z",
    "2025-11-27T14:39:01.663Z",
    "2025-11-27T14:39:00.363Z",
    "2025-11-27T14:38:58.963Z",
    "2025-11-27T14:38:57.563Z",
    "2025-11-27T14:38:56.263Z",
    "2025-11-27T14:38:54.963Z",
    "2025-11-27T14:38:53.663Z",
    "2025-11-27T14:38:52.363Z",
    "2025-11-27T14:38:50.963Z",
    "2025-11-27T14:38:49.663Z",
    "2025-11-27T14:38:48.363Z",
    "2025-11-27T14:38:47.063Z",
    "2025-11-27T14:38:45.763Z",
    "2025-11-27T14:38:44.463Z",
    "2025-11-27T14:38:43.063Z",
    "2025-11-27T14:38:41.763Z",
    "2025-11-27T14:38:40.463Z",
    "2025-11-27T14:38:39.163Z",
    "2025-11-27T14:38:37.763Z",
    "2025-11-27T14:38:36.463Z",
    "2025-11-27T14:38:35.163Z",
    "2025-11-27T14:37:33.263Z",
    "2025-11-27T14:36:30.363Z",
    "2025-11-27T14:33:40.963Z",
    "2025-11-27T14:31:52.363Z",
    "2025-11-27T14:31:41.563Z",
    "2025-11-27T14:27:46.563Z",
    "2025-11-27T14:20:08.217Z",
    "2025-11-27T14:20:08.217Z",
    "2025-11-27T14:20:08.207Z",
    "2025-11-27T14:20:08.195Z",
    "2025-11-27T14:20:08.185Z",
    "2025-11-27T14:20:08.185Z",
    "2025-11-27T14:20:08.167Z",
    "2025-11-27T14:20:08.167Z",
    "2025-11-27T13:47:17.565Z",
    "2025-11-27T13:47:17.557Z",
    "2025-11-27T13:47:17.548Z",
    "2025-11-27T13:47:17.545Z",
    "2025-11-27T13:47:17.543Z",
    "2025-11-27T13:47:17.533Z",
    "2025-11-27T13:47:17.533Z",
    "2025-11-27T13:47:17.525Z",
    "2025-11-27T13:47:17.520Z",
    "2025-11-27T13:47:17.518Z",
    "2025-11-27T13:47:17.513Z",
    "2025-11-27T13:47:17.513Z",
    "2025-11-27T13:47:17.513Z",
    "2025-11-27T13:45:44.567Z",
    "2025-11-27T13:45:44.567Z",
    "2025-11-27T13:45:44.557Z",
    "2025-11-27T13:45:44.555Z",
    "2025-11-27T13:45:44.553Z",
    "2025-11-27T13:45:44.548Z",
    "2025-11-27T13:45:44.545Z",
    "2025-11-27T13:45:44.543Z",
    "2025-11-27T13:45:44.543Z",
    "2025-11-27T13:45:44.525Z",
    "2025-11-27T13:45:44.513Z",
    "2025-11-27T13:45:44.513Z"
]
```

`processingUpdatedAt` is `'2025-11-27T14:20:52.819Z'`

You see most of the docs in the sample are newer than the edit time, but not all of them, some old ones stick around. The thing is, there _are_ many many docs that are newer, it's just that the data source doesn't return them:
```
POST _query
{
  "query": "FROM logs.android | WHERE @timestamp > \"2025-11-27T14:20:52.819Z\" | STATS COUNT(*)"
}
```

returns 653 matches.

I think the problem is that it the search uses `terminate_after`:

```
{
  "index": "logs.windows",
  "allow_no_indices": true,
  "query": {
    "bool": {
      "must": [],
      "filter": [],
      "should": [],
      "must_not": []
    }
  },
  "sort": [
    {
      "@timestamp": {
        "order": "desc"
      }
    }
  ],
  "size": 100,
  "terminate_after": 100,
  "track_total_hits": false
}
```

Could you look into that? Not sure what the right approach would be here - the `terminate_after` makes the request return fast, but on the flip side, it's not necessarily the most recent 100 docs, just the first 100 it found in an ordered fashion.

Maybe having terminate_after 1000 or so could do the trick.

AlejandroFrndz committed:
fix scout

AlejandroFrndz committed:
remove terminate after

AlejandroFrndz committed:
Adds telemetry for simulation sample fetch latency

Adds telemetry to measure latency of fetching simulation samples during stream simulation.

Integrates the telemetry client into the data collection flow so timing starts when the search subscription begins and reports duration when it finalizes. Registers a new telemetry event and schema that capture stream_type, data_source_type, and duration_ms. Enables tracking of sample-fetch performance to support monitoring and future optimizations.

AlejandroFrndz committed:
fix types

AlejandroFrndz committed:
Adds search timeout and error handling

Adds a 10s timeout to data source searches and surfaces a dedicated timeout toast. Applies the timeout before result validation/mapping so long-running requests are aborted promptly. Uses a formatted error for non-timeout failures to keep messages consistent and provide clearer user guidance when fetches exceed the time limit.

AlejandroFrndz committed:
Adds stream name to simulation fetch latency telemetry

Adds the stream name to simulation samples fetch latency telemetry so fetch latency can be attributed to specific streams. Updates the telemetry schema, types, client API surface, and the caller to include and pass the stream name when starting latency tracking.

AlejandroFrndz committed:
fix jest tests

flash1293 reviewed (approved):
Looks mostly good to me, thanks! Just one small-ish comment.

However, since this is a big PR I would appreciate another pair of eyes on it - @tonyghiani since you are very familiar with the processing code, could you take another look?

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/368211)
* Commit: 1148b8281347af84e6ea580adda1386b118ae271
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-242904)

### Failed CI Steps
* [FTR Configs #5](https://buildkite.com/elastic/kibana-pull-request/builds/368211#019ae3c0-ec6c-419a-bc80-d2540698c956)
* [FTR Configs #139](https://buildkite.com/elastic/kibana-pull-request/builds/368211#019ae3c0-ed0e-4c4e-9744-2e916ca1511d)
* [FTR Configs #139](https://buildkite.com/elastic/kibana-pull-request/builds/368211#019ae3ed-c7d1-4c9e-9ce5-730cdaabb89d)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/368211#019ae3ed-c7d1-4c9e-9ce5-730cdaabb89d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/368211/jobs/019ae3ed-c7d1-4c9e-9ce5-730cdaabb89d/artifacts/019ae3fc-ec49-48ba-8276-eee4426148c2) FTR Configs #<span></span>139 / Agent Builder agents Edit agent should edit agent name
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/368211#019ae3c0-ed0e-4c4e-9744-2e916ca1511d) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/368211/jobs/019ae3c0-ed0e-4c4e-9744-2e916ca1511d/artifacts/019ae3ed-63cb-4f2f-bf20-37bd50ab3d7b) FTR Configs #<span></span>139 / Agent Builder converse Conversation Error Handling can start a new conversation when there is an error
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/368211#019ae3c0-ec6c-419a-bc80-d2540698c956) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/368211/jobs/019ae3c0-ec6c-419a-bc80-d2540698c956/artifacts/019ae3eb-da32-41dc-ac62-2e685170c715) FTR Configs #<span></span>5 / InfraOps App Metrics UI Home page with metrics present renders the kubernetes tour component and allows user to dismiss it without seeing it again

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/cace26111244608777cfddc23cd9a674b39afeee) | [after](https://github.com/elastic/kibana/commit/1148b8281347af84e6ea580adda1386b118ae271) | diff |
| --- | --- | --- | --- |
| `datasetQuality` | 963 | 965 | +2 |
| `streamsApp` | 1106 | 1108 | +2 |
| total |  |  | +4 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/cace26111244608777cfddc23cd9a674b39afeee) | [after](https://github.com/elastic/kibana/commit/1148b8281347af84e6ea580adda1386b118ae271) | diff |
| --- | --- | --- | --- |
| `@kbn/streams-schema` | 187 | 189 | +2 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/cace26111244608777cfddc23cd9a674b39afeee) | [after](https://github.com/elastic/kibana/commit/1148b8281347af84e6ea580adda1386b118ae271) | diff |
| --- | --- | --- | --- |
| `datasetQuality` | 499.4KB | 500.2KB | +791.0B |
| `streamsApp` | 1.1MB | 1.1MB | +3.1KB |
| total |  |  | +3.9KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/cace26111244608777cfddc23cd9a674b39afeee) | [after](https://github.com/elastic/kibana/commit/1148b8281347af84e6ea580adda1386b118ae271) | diff |
| --- | --- | --- | --- |
| `@kbn/streams-schema` | 19 | 20 | +1 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/cace26111244608777cfddc23cd9a674b39afeee) | [after](https://github.com/elastic/kibana/commit/1148b8281347af84e6ea580adda1386b118ae271) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 21.3KB | 22.0KB | +685.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/cace26111244608777cfddc23cd9a674b39afeee) | [after](https://github.com/elastic/kibana/commit/1148b8281347af84e6ea580adda1386b118ae271) | diff |
| --- | --- | --- | --- |
| `@kbn/streams-schema` | 187 | 194 | +7 |


</details>


### History
* :broken_heart: [Build #368104](https://buildkite.com/elastic/kibana-pull-request/builds/368104) failed c4a7b865ddc9996ef56e10fe8fa8d167367d1a0e
* :broken_heart: [Build #367924](https://buildkite.com/elastic/kibana-pull-request/builds/367924) failed c4a7b865ddc9996ef56e10fe8fa8d167367d1a0e
* :broken_heart: [Build #367822](https://buildkite.com/elastic/kibana-pull-request/builds/367822) failed d719af7d5d14fbc20a3b2fcf2cedffc67a0421c7
* :green_heart: [Build #367115](https://buildkite.com/elastic/kibana-pull-request/builds/367115) succeeded 318b2f8d6f2d78295c5e1f7a3021c1b71619086b
* :broken_heart: [Build #366737](https://buildkite.com/elastic/kibana-pull-request/builds/366737) failed a3ac625a1f76c6f53fc43cf716d4efd5a0105937

cc @AlejandroFrndz

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368211","number":368211,"commit":"1148b8281347af84e6ea580adda1386b118ae271"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368104","number":368104,"commit":"c4a7b865ddc9996ef56e10fe8fa8d167367d1a0e"},{"buildStatus":{"state":"running","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367924","number":367924,"commit":"c4a7b865ddc9996ef56e10fe8fa8d167367d1a0e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367822","number":367822,"commit":"d719af7d5d14fbc20a3b2fcf2cedffc67a0421c7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367115","number":367115,"commit":"318b2f8d6f2d78295c5e1f7a3021c1b71619086b"}],"number":368211}
buildkite-pr-comment-->

elasticmachine commented:
Pinging @elastic/obs-presentation-team (Team:obs-presentation)

tonyghiani commented on x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/streams/group_stream.ts:
```diff
@@ -77,10 +77,11 @@ export class GroupStream extends StreamActiveRecord<Streams.GroupStream.Definiti
             classic: {},
             lifecycle: { inherit: {} },
-            processing: { steps: [] },
+            processing: { steps: [], updated_at: new Date().toISOString() },
```
Same here as in https://github.com/elastic/kibana/pull/242904/files#r2585201629, please also check it anywhere else in the PR

tonyghiani reviewed (commented):
Just left a couple notes, nothing critical though 👌

AlejandroFrndz commented on x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/state_management/data_source_state_machine/data_collector_actor.ts:
```diff
@@ -71,15 +94,36 @@ function getDataCollectorForDataSource(dataSource: EnrichmentDataSourceWithUIAtt
+            registerFetchLatency();
+          },
+        }),
+        timeout(SEARCH_TIMEOUT_MS),
```
How do we wanna handle user feedback for ES side timeouts? Because right now with the client-side timeout we display the error toast as we would not have any data but now that we might have partial data... 

I'm thinking if the request times out without even giving partial data we can error much like right now on the client side, and display the toast as usual. But, if the query timed out but still managed to get some partial data back. Should we let users know somehow, as there would be less than the usual 100 samples?

flash1293 replied:
Just showing the partial data is the default behavior, and I think it's OK here. We don't need to show a toast or anything because it's not an unexpected behavior and it's not really actionable anyway.

AlejandroFrndz committed:
Unifies updated_at timestamps per operation

Captures a single ISO timestamp at the start of each operation and reuses it for all updated_at fields across related stream definitions and processing metadata. This ensures consistency of timestamps for cascading/upsert actions and avoids tiny differences from multiple Date calls.

Also reduces repeated Date allocations for a small performance improvement; no functional behavior change beyond timestamp consistency.

AlejandroFrndz committed:
Treats unknown data source types as not outdated

Prevent runtime errors for unrecognized data source types by returning false instead of throwing an exception.

This simplifies the branching logic by removing an explicit case and making unknown or future data source types be considered not outdated, improving robustness and forward compatibility.

AlejandroFrndz committed:
Uses data source selector for context

Replace direct snapshot access with a selector hook to obtain the active data source context.

This makes outdated-document detection reactive to data source context changes and avoids stale snapshot reads that could cause incorrect "outdated" state.

AlejandroFrndz committed:
Uses ES query timeout and simplifies error handling

Replaces the RxJS-based search timeout with an Elasticsearch query timeout and removes the RxJS timeout operator to let the backend enforce request time limits. Unifies error formatting and notification logic by using a single formatted error for all failures instead of a special-case timeout toast.

This reduces duplicated timeout logic, simplifies imports and constants, and produces more consistent user-facing error messages.

AlejandroFrndz committed:
fix tests

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/368617)
* Commit: d09ea4be0059fde7c28825727dd3f6bf24591f7b
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-242904)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/3db1487e73c81093381e523f098e9727b8912e60) | [after](https://github.com/elastic/kibana/commit/d09ea4be0059fde7c28825727dd3f6bf24591f7b) | diff |
| --- | --- | --- | --- |
| `datasetQuality` | 964 | 966 | +2 |
| `streamsApp` | 1114 | 1116 | +2 |
| total |  |  | +4 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/3db1487e73c81093381e523f098e9727b8912e60) | [after](https://github.com/elastic/kibana/commit/d09ea4be0059fde7c28825727dd3f6bf24591f7b) | diff |
| --- | --- | --- | --- |
| `@kbn/streams-schema` | 217 | 219 | +2 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/3db1487e73c81093381e523f098e9727b8912e60) | [after](https://github.com/elastic/kibana/commit/d09ea4be0059fde7c28825727dd3f6bf24591f7b) | diff |
| --- | --- | --- | --- |
| `datasetQuality` | 504.6KB | 505.4KB | +794.0B |
| `streamsApp` | 1.1MB | 1.1MB | +2.6KB |
| total |  |  | +3.4KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/3db1487e73c81093381e523f098e9727b8912e60) | [after](https://github.com/elastic/kibana/commit/d09ea4be0059fde7c28825727dd3f6bf24591f7b) | diff |
| --- | --- | --- | --- |
| `@kbn/streams-schema` | 19 | 20 | +1 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/3db1487e73c81093381e523f098e9727b8912e60) | [after](https://github.com/elastic/kibana/commit/d09ea4be0059fde7c28825727dd3f6bf24591f7b) | diff |
| --- | --- | --- | --- |
| `streamsApp` | 21.3KB | 22.0KB | +685.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/3db1487e73c81093381e523f098e9727b8912e60) | [after](https://github.com/elastic/kibana/commit/d09ea4be0059fde7c28825727dd3f6bf24591f7b) | diff |
| --- | --- | --- | --- |
| `@kbn/streams-schema` | 219 | 226 | +7 |


</details>


### History
* :broken_heart: [Build #368572](https://buildkite.com/elastic/kibana-pull-request/builds/368572) failed fc9fcabf699f223d396b5871f935a89fe40e8ba6
* :broken_heart: [Build #368211](https://buildkite.com/elastic/kibana-pull-request/builds/368211) failed 1148b8281347af84e6ea580adda1386b118ae271
* :broken_heart: [Build #368104](https://buildkite.com/elastic/kibana-pull-request/builds/368104) failed c4a7b865ddc9996ef56e10fe8fa8d167367d1a0e
* :broken_heart: [Build #367924](https://buildkite.com/elastic/kibana-pull-request/builds/367924) failed c4a7b865ddc9996ef56e10fe8fa8d167367d1a0e
* :broken_heart: [Build #367822](https://buildkite.com/elastic/kibana-pull-request/builds/367822) failed d719af7d5d14fbc20a3b2fcf2cedffc67a0421c7

cc @AlejandroFrndz

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368617","number":368617,"commit":"d09ea4be0059fde7c28825727dd3f6bf24591f7b"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368572","number":368572,"commit":"fc9fcabf699f223d396b5871f935a89fe40e8ba6"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368211","number":368211,"commit":"1148b8281347af84e6ea580adda1386b118ae271"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368104","number":368104,"commit":"c4a7b865ddc9996ef56e10fe8fa8d167367d1a0e"},{"buildStatus":{"state":"running","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367924","number":367924,"commit":"c4a7b865ddc9996ef56e10fe8fa8d167367d1a0e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367822","number":367822,"commit":"d719af7d5d14fbc20a3b2fcf2cedffc67a0421c7"}],"number":368617}
buildkite-pr-comment-->
